File: golang/src/detectors/thread-safety-violation/ThreadSafetyViolation.go
package main

import (
	"sync"
)

// {fact rule=thread-safety-violation@v1.0 defects=1}
func threadSafetyViolationNoncompliant(){
    var counter int
    var wg sync.WaitGroup

    for i:=0; i < 1000; i++{
        wg.Add(1)
        go func() {
            // Noncompliant: incrementing without synchronization.
            counter ++
            wg.Done()
        }()
    }
    wg.Wait()
}
// {/fact}

// {fact rule=thread-safety-violation@v1.0 defects=0}
func threadSafetyViolationCompliant(){
    var counter int
    var wg sync.WaitGroup
    var mu sync.Mutex
    for i:=0; i < 1000; i++{
        wg.Add(1)
        go func() {
            mu.Lock()
            // Compliant: `sync.Mutex` is used to ensure only one goroutine can modify shared data at a time.
            counter ++
            mu.Unlock()
            wg.Done()
        }()
    }
    wg.Wait()
}
// {/fact}

File: csharp/src/detectors/thread-safety-violation/ThreadSafetyViolation.cs
namespace ThreadSafe
{
    class ThreadSafetyViolation
    {
        int result = 0;
        private readonly object mainLock = new object();
        
        // {fact rule=thread-safety-violation@v1.0 defects=1}
        public void ThreadSafetyViolationNoncompliant()
        {
            for (int i = 0; i < 100000; i++)
            {
                // Noncompliant: This global variable is accessed without a lock.
                result = result + 1;
            }
        }
        // {/fact}
        
        // {fact rule=thread-safety-violation@v1.0 defects=0}
        public void ThreadSafetyViolationCompliant()
        {
            for (int i = 0; i < 100000; i++)
            {
                // Compliant: Provide lock to protect the integrity of the global variable.
                Monitor.Enter(mainLock);
                result = result + 1;
                Monitor.Exit(mainLock);
            }
        }
        // {/fact}
    }
}

